src/pybind/mgr/rgw/module.py

```diff
@@ -99,25 +102,75 @@ def _cmd_rgw_admin(self, params: Sequence[str]):
+                                                               endpoints,
+                                                               sys_uid,
+                                                               uid,
+                                                               start_radosgw,
+                                                               zonegroup_is_master,
+                                                               zone_is_master)
         except RGWAMException as e:
             self.log.error('cmd run exception: (%d) %s' % (e.retcode, e.message))
             return (e.retcode, e.message, e.stderr)
 
         return HandleCommandResult(retval=retval, stdout=out, stderr=err)
 
+    @CLICommand('rgw apply', perm='rw')
+    def _cmd_rgw_apply(self, inbuf: Optional[str] = None):
+        """apply rgw spec"""
+
+        usage = """Usage: ceph rgw apply -i <yaml spec>"""
+        if not inbuf:
+            return HandleCommandResult(retval=-errno.EINVAL, stdout='', stderr=usage)
+
+        # YAML '---' document separator with no content generates
+        # None entries in the output. Let's skip them silently.
+        yaml_objs: Iterator = yaml.safe_load_all(inbuf)
+        specs = [o for o in yaml_objs if o is not None]
+        outs = []
+        err = ''
+        retval = 0
+        for spec in specs:
+            try:
+                realm_name = spec.get('realm_name')
+                zonegroup_name = spec.get('zonegroup_name')
+                zone_name = spec.get('zone_name')
+                if not all([realm_name, zonegroup_name, zone_name]):
+                    return HandleCommandResult(retval=-errno.EINVAL, stdout='', stderr='spec fields: realm_name, zonegroup_name and zone_name are mandatory')
+                retval, out, err = RGWAM(self.env).realm_bootstrap(realm_name,
+                                                                   zonegroup_name,
+                                                                   zone_name,
+                                                                   spec.get('endpoints'),
+                                                                   spec.get('sys_uid'),
+                                                                   spec.get('uid'),
+                                                                   spec.get('start_radosgw'),
+                                                                   spec.get('zonegroup_is_master'),
+                                                                   spec.get('zone_is_master'))
+                if retval == 0:
+                    outs.append(f'{out}')
+                else:
+                    break
+            except RGWAMException as e:
+                self.log.error(f'cmd run exception: {e.retcode} {e.message}')
+                retval = e.retcode
+                err = f'{e.message}'
+                break
+
+        return HandleCommandResult(retval=retval, stdout=f'{outs}', stderr=f'{err}')
```

Comment:
If I understand the code correctly, `outs` is a list of strings and thus the stdout here will be a repr of the list (like: `["foo", "ar"]`)
I suggest using JSON formatting instead of relying on a python repr. Especially since this is a  net-new command there's no backwards compatibility to worry about so we should start off with standardized structured output. JSON is more standard then python repr.   (It would also make a future conversion to the new decorators I'm working on easier. hint hint :-) )

Reply:
TBH I wasn't sure which output format to use (so I used python repr) but your arguments are solid enough to change it to JSON. I'll be fixing it :+1: